twenty-first/src/math/polynomial.rs

```diff
@@ -1114,32 +1185,38 @@ where
                 [vec![FF::from(0); n], vec![self.coefficients[0].inverse()]].concat(),
             );
         }
+
         let reverse = self.reverse();
+
         // The next function gives back a polynomial g(X) of degree at most arg,
         // such that f(X) * g(X) = 1 mod X^arg.
         // Without modular reduction, the degree of the product f(X) * g(X) is
         // deg(f) + arg -- even after coefficient reversal. So n = deg(f) + arg
         // and arg = n - deg(f).
         let inverse_reverse = reverse.formal_power_series_inverse_minimal(n - degree);
         let product_reverse = reverse.multiply(&inverse_reverse);
-        product_reverse.reverse()
+        let product = product_reverse.reverse();
+
+        // Coefficient reversal drops trailing zero. Correct for that.
+        let product_degree = product.degree() as usize;
+        product.shift_coefficients(n - product_degree)
     }
 
     /// Reduces f(X) by a structured modulus, which is of the form
     /// X^{m+n} + (something of degree less than m). When the modulus has this
     /// form, polynomial modular reductions can be computed faster than in the
     /// generic case.
     fn reduce_by_structured_modulus(&self, multiple: &Self) -> Self {
-        let leading_term = Polynomial::new(
-            [
-                vec![FF::from(0); multiple.degree() as usize],
-                vec![FF::from(1); 1],
-            ]
-            .concat(),
-        );
+        assert_ne!(multiple.degree(), 0);
+        let multiple_degree =
+            usize::try_from(multiple.degree()).unwrap_or_else(|_| panic!("cannot reduce by zero"));
```

Comment:
Without a `format!` in the error message, this can be:

```suggestion
            usize::try_from(multiple.degree()).expect("cannot reduce by zero");
```